src/afa/afa.cc

```diff
@@ -34,29 +36,531 @@ const std::string Mata::Afa::TYPE_AFA = "AFA";
+            {
+                result.insert(el);
+            }
+        }
+    }
+    return ClosedSet<State>(ClosedSet<State>::downward_closed, 0, transRelation.size()-1, result);
+} // pre }}}
+
+/** This function takes set of nodes and a symbol and returns all the nodes which are able to access the given nodes
+* in one step using the given symbol. The output will be represented as ClosedSet to omit the neccessity
+* to explicitly return all the proper nodes
+* The result is always a downward closed set!
+* @param nodes source nodes
+* @param symb a symbol used to perform an inverse transition
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(Nodes nodes, Symbol symb) const
+{
+    ClosedSet<State> result(ClosedSet<State>::downward_closed, 0, transRelation.size()-1);
+    for(auto node : nodes)
+    {
+        result = result.Union(pre(node, symb));
+    }
+    return result;
+} // pre }}}
+
+/** This function performs pre directly over the given closed set and the given symbol
+* The result is always a downward closed set!
+* @param closed_set a closed set
+* @param symb a symbol used to perform an inverse transition
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(ClosedSet<State> closed_set, Symbol symb) const
+{
+    assert(closed_set.get_type() == ClosedSet<State>::downward_closed);
+    return pre(closed_set.get_antichain(), symb);
+} // pre }}}
+
+/** This function allows us to perfom pre over a node and use the whole alphabet to perform individual transitions 
+* The result is always a downward closed set!
+* @param node a set of states
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(Node node) const
+{
+    if(node.empty())
+    {
+        return ClosedSet<State>(ClosedSet<State>::downward_closed, 0, transRelation.size()-1, Nodes{Node{}});
+    }
+    ClosedSet<State> result(ClosedSet<State>::downward_closed, 0, transRelation.size()-1);
+    for(auto transVec : invTransRelation[*(node.begin())])
+    {
+        result.insert(pre(node, transVec.symb).get_antichain());
+    }
+    return result;
+} // pre }}}
+
+/** This function allows us to perfom pre over a set of nodes and use the whole alphabet to perform individual transitions 
+* The result is always a downward closed set!
+* @param nodes a set of sets of states
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(Nodes nodes) const
+{
+    ClosedSet<State> result(ClosedSet<State>::downward_closed, 0, transRelation.size()-1);
+    for(auto node : nodes)
+    {
+        result.insert(pre(node).get_antichain());
+    }
+    return result;
+} // pre }}}
+
+// The end of the definitions of the "pre" functions
+////////////////////////////////////////////////////
+
 bool Afa::has_trans(const Trans& trans) const
 { // {{{
-  return std::find(this->transitions.begin(),
-                   this->transitions.end(),
-                   trans) != this->transitions.end();
+    Nodes res = *perform_trans(trans.src, trans.symb);
```

Comment:
Is this really the efficient process? It seams shady that you perform the transition to check if the automaton has the transition. Can you elaborate?

Reply:
Thanks for your suggestion, I have reimplemented it.